src/herder/TxSetFrame.cpp

```diff
@@ -521,17 +507,11 @@ TxSetFrame::checkValid(Application& app, uint64_t lowerBoundCloseTimeOffset,
         return false;
     }
 
-    if (!std::is_sorted(mTxs.begin(), mTxs.end(), &TxSetUtils::hashTxSorter))
-    {
-        CLOG_DEBUG(Herder, "Got bad txSet: {} is not in hash order",
-                   hexAbbrev(mPreviousLedgerHash));
-        return false;
-    }
-
-    if (std::adjacent_find(mTxs.begin(), mTxs.end(),
+    auto sortedTxs = TxSetUtils::sortTxsInHashOrder(mTxs);
```

Comment:
> for a txset to be valid, transactions (in a group in the case of GeneralizedTxSets) are supposed to be sorted in hash order.

This is a condition defined by the protocol and it's applicable to XDR only, the same way as e.g. the component order for generalized tx set.  I would argue that in-memory representation should not worry about the XDR representation (besides the (de-)serialization boundaries), which is why I wanted to signify that by not doing anything specific with in-memory representation (btw even with your suggestion `mTxs` would have somewhat arbitrary order for generalized tx set). There really was never a need to mix the on-the-wire tx ordering with anything we did with tx set frames.

> This check here makes it look like the only thing we care about is that we don't have duplicates.
> I would just remove this check and ensure that txsets follow proper invariants at construction time.

We cannot remove this check, as the main purpose of `checkValid` is to validate external tx sets. If you suggest moving this check to XDR structure validation, then I'm not sure why this check among the others. It's possible to do it on raw XDR, but the code would be even more involved, as we would need to merge all the txs in generalized tx set and basically duplicate the logic we already have.

The main purpose of `checkValid` is to make sure that *external* tx set is valid, right? The internal call is there to just make sure we don't have a bug. I'm not opposed to doing an XDR roundtrip for nominated tx sets to be more paranoid, but that wouldn't allow us to remove any checks from `checkValid`.

In general, I want `validateTxSetXDRStructure` to perform the checks that only are needed for the serialized format (component order, tx order etc) and `checkValid` perform the checks that are independent of the serialized format. I realize that some of these checks (such as duplicates) could belong to either method and for those we just put it where it's more convenient to perform the check.

Reply:
> We cannot remove this check, as the main purpose of checkValid is to validate external tx sets. If you suggest moving this check to XDR structure validation, then I'm not sure why this check among the others. It's possible to do it on raw XDR, but the code would be even more involved, as we would need to merge all the txs in generalized tx set and basically duplicate the logic we already have.

Yes, I was suggesting moving this to the factory functions yeah as `TxSet` is supposed to be immutable (in terms of composition). I don't know if performing the check on the xdr is necessary, you could also just make `mTxs` a `std::set` and call it a day?

> The main purpose of `checkValid` is to make sure that external tx set is valid, right? The internal call is there to just make sure we don't have a bug. I'm not opposed to doing an XDR roundtrip for nominated tx sets to be more paranoid, but that wouldn't allow us to remove any checks from `checkValid`.

yeah as `GeneralizedTransactionSet` allows for pretty complicated things (groups and phases), I would much prefer to do the most paranoid check: it would have caught the issue we had recently, and will catch potential other issues related to other fields. I do understand that this should "never happen" and be covered by the serialization/de-serialization unit tests.

Reply:
Sorry, I realized that what I wrote is quite ambiguous.

This PR touches on a couple things, and I think we've been talking about different things. Let me try to clarify!

Reason I think it's important is that I don't want to keep revisiting this as I am pretty sure the internal representation is going to continue to evolve over the next few weeks/months.

First, that the in-memory representation should not be coupled with the on wire format. I agree that this became quite muddy with the recent refactors, and making this not ambiguous makes perfect sense to me.
The internal representation should optimize for ease of maintenance and use cases of the different methods.
As far as giving access to that internal structure via a simple `std::vector`, I am not sure this will hold much longer with the Jump Cannon work, but we'll see later (a visitor pattern may hold better).

Second, is around `checkValid`. In the original issue (and my comment above), I was drawing the distinction between invariants at the class level vs what a method like `checkValid` does.
When we made `TxSetFrame` immutable, this really allowed us to separate concerns between conditions that must be true at any time (for whatever reason) vs additional conditions that depend on outside components (like ledger state).
In addition to separation of concerns, we also get "for free" the potential performance benefits of not performing redundant checks (`checkValid` is called multiple times). Those are similar gains that we got when you introduced `InvalidTxSetFrame`.
Anyways: `checkValid` already depends on the internal structures being in a good state (like fee map initialized, txs cannot be `nullptr`, etc), so expecting it to contain a unique set of transactions seems to line up with this kind of expectation.
I am pretty sure we would not have this conversation at all if the internal representation was an `std::set` instead of a `std::vector` (it's called `TxSet`, not `TxMultiSet` or whatever) as we would not be able to represent using the "in-memory" representation, whatever comes from the wire/other classes in the factory methods.
Now, should we use `std::set` (or `std::unordered_set`) here? I am not sure: going back to the "use cases", we only construct things once and we never lookup by hash. Sorting by hash and looking for duplicates is a perfect way to do it, all I am arguing is to leave this as a class invariant (enforced at construction time).

Hopefully this allows to move forward, it feels we're bike-shedding a bit on this.

Reply:
Ok, thanks for the explanation. I've added the XDR roundtrip and removed the duplicates check as it's covered by seq num checks.

Reply:
cool change looks good now - I am not sure why you have some tests failing though

Reply:
Some tests relied on tx pointers remaining stable during the ledger close, I've updated them to use the result meta (which I believe is a correct thing to do anyway).